File: Problems/Algorithms/823. Binary Trees with Factors/FactorBinaryTree.java
class Solution {
    public int numFactoredBinaryTrees(int[] arr) {
        int MOD = 1_000_000_007;
        
        int n = arr.length;
        Arrays.sort(arr);
        
        long[] dp = new long[n];
        Arrays.fill(dp, 1);
        
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map.put(arr[i], i);
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[i] % arr[j] == 0) {
                    int key = arr[i] / arr[j];
                    if (map.containsKey(key)) {
                        dp[i] = (dp[i] + dp[j] * dp[map.get(key)]) % MOD;
                    }
                }
            }
        }
        
        long ans = 0;
        for (long num : dp) {
            ans += num;
        }
        return (int) (ans % MOD);
    }
}

File: Problems/Algorithms/823. Binary Trees with Factors/factor_binary_tree.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        mod = 10 ** 9 + 7

        arr.sort()
        factors = set(arr)
        dp = {num: 1 for num in arr}

        for i in arr:
            for j in arr:
                if j > i ** 0.5:
                    break
                
                if i % j == 0 and i // j in factors:
                    if i // j == j:
                        dp[i] += dp[j] * dp[j]
                    else:
                        dp[i] += dp[j] * dp[i // j] * 2
                    dp[i] %= mod

        return sum(dp.values()) % mod
